include/ensmallen_bits/problems/logistic_regression_function_impl.hpp

```diff
@@ -129,97 +135,100 @@ typename MatType::elem_type LogisticRegressionFunction<MatType>::Evaluate(
+
   // Regularization term.
   MatType regularization;
   regularization = lambda * parameters.tail_cols(parameters.n_elem - 1);
 
-  const arma::Row<ElemType> sigmoids = (1 / (1 + arma::exp(-parameters(0, 0)
+  const BaseRowType sigmoids = (1 / (1 + exp(-parameters(0, 0)
       - parameters.tail_cols(parameters.n_elem - 1) * predictors)));
 
-  gradient.set_size(arma::size(parameters));
-  gradient[0] = -arma::accu(responses - sigmoids);
+  gradient.set_size(size(parameters));
+  gradient[0] = -accu(responses - sigmoids);
   gradient.tail_cols(parameters.n_elem - 1) = (sigmoids - responses) *
       predictors.t() + regularization;
 }
 
 //! Evaluate the gradient of the logistic regression objective function for a
 //! given batch size.
-template<typename MatType>
+template<typename MatType, typename LabelsType>
 template<typename GradType>
-void LogisticRegressionFunction<MatType>::Gradient(
+void LogisticRegressionFunction<MatType, LabelsType>::Gradient(
                 const MatType& parameters,
                 const size_t begin,
                 GradType& gradient,
                 const size_t batchSize) const
 {
-  typedef typename MatType::elem_type ElemType;
+  typedef typename ForwardType<MatType>::brow BaseRowType;
 
   // Regularization term.
   MatType regularization;
   regularization = lambda * parameters.tail_cols(parameters.n_elem - 1)
       / predictors.n_cols * batchSize;
 
-  const arma::Row<ElemType> exponents = parameters(0, 0) +
+  const BaseRowType exponents = parameters(0, 0) +
       parameters.tail_cols(parameters.n_elem - 1) *
       predictors.cols(begin, begin + batchSize - 1);
   // Calculating the sigmoid function values.
-  const arma::Row<ElemType> sigmoids = 1.0 / (1.0 + arma::exp(-exponents));
+  const BaseRowType sigmoids = 1.0 / (1.0 + exp(-exponents));
 
   gradient.set_size(parameters.n_rows, parameters.n_cols);
-  gradient[0] = -arma::accu(responses.subvec(begin, begin + batchSize - 1) -
+  gradient[0] = -accu(conv_to<BaseRowType>::from(responses.subvec(begin, begin + batchSize - 1)) -
       sigmoids);
   gradient.tail_cols(parameters.n_elem - 1) = (sigmoids -
-      responses.subvec(begin, begin + batchSize - 1)) *
+      conv_to<BaseRowType>::from(responses.subvec(begin, begin + batchSize - 1))) *
       predictors.cols(begin, begin + batchSize - 1).t() + regularization;
 }
 
 /**
  * Evaluate the partial gradient of the logistic regression objective
  * function with respect to the individual features in the parameter.
  */
-template <typename MatType>
-void LogisticRegressionFunction<MatType>::PartialGradient(
+template <typename MatType, typename LabelsType>
+template<typename ElemType>
+void LogisticRegressionFunction<MatType, LabelsType>::PartialGradient(
     const MatType& parameters,
     const size_t j,
-    arma::sp_mat& gradient) const
+    arma::SpMat<ElemType>& gradient) const
 {
   const arma::Row<typename MatType::elem_type> diffs = responses -
```

Comment:
I guess you can't use the implementation below because the result here has to be dense?